crates/lib/src/qpu/api.rs

```diff
@@ -91,74 +74,316 @@ impl From<String> for JobId {
+                get_controller_job_results_request::Target::EndpointId(endpoint_id.to_string()),
+            ),
+            Self::Gateway | Self::DirectAccess => Ok(
+                get_controller_job_results_request::Target::QuantumProcessorId(
+                    quantum_processor_id
+                        .ok_or(QpuApiError::MissingQpuId)?
+                        .to_string(),
+                ),
+            ),
+        }
+    }
+
+    async fn get_controller_client(
+        &self,
+        client: &Qcs,
+        quantum_processor_id: Option<&str>,
+    ) -> Result<ControllerClient<RefreshService<Channel, ClientConfiguration>>, QpuApiError> {
+        let address = match self {
+            Self::EndpointId(endpoint_id) => {
+                let endpoint = get_endpoint(&client.get_openapi_client(), endpoint_id).await?;
+                endpoint
+                    .addresses
+                    .grpc
+                    .ok_or_else(|| QpuApiError::EndpointNotFound(endpoint_id.into()))?
+            }
+            Self::Gateway => {
+                self.get_gateway_address(
+                    quantum_processor_id.ok_or(QpuApiError::MissingQpuId)?,
+                    client,
+                )
+                .await?
+            }
+            Self::DirectAccess => {
+                self.get_default_endpoint_address(
+                    quantum_processor_id.ok_or(QpuApiError::MissingQpuId)?,
+                    client,
+                )
+                .await?
+            }
+        };
+        Self::grpc_address_to_client(&address, client)
+    }
+
+    fn grpc_address_to_client(
+        address: &str,
+        client: &Qcs,
+    ) -> Result<ControllerClient<RefreshService<Channel, ClientConfiguration>>, QpuApiError> {
+        let uri = parse_uri(address).map_err(QpuApiError::GrpcError)?;
+        let channel = get_channel(uri).map_err(|err| QpuApiError::GrpcError(err.into()))?;
+        let service = wrap_channel_with(channel, client.get_config().clone());
+        Ok(ControllerClient::new(service)
+            .max_decoding_message_size(MAX_DECODING_MESSAGE_SIZE_BYTES))
+    }
+
+    async fn get_gateway_address(
+        &self,
+        quantum_processor_id: &str,
+        client: &Qcs,
+    ) -> Result<String, QpuApiError> {
+        let mut min = None;
+        let mut next_page_token = None;
+        loop {
+            let accessors = list_quantum_processor_accessors(
+                &client.get_openapi_client(),
+                quantum_processor_id,
+                Some(100),
+                next_page_token.as_deref(),
+            )
+            .await?;
+
+            let accessor = accessors
+                .accessors
+                .into_iter()
+                .filter(|acc| {
+                    acc.live
+                    // `as_deref` needed to work around the `Option<Box<_>>` type.
+                    && acc.access_type.as_deref() == Some(&QuantumProcessorAccessorType::GatewayV1)
```

Comment:
👍🏻 

I definitely thought there would be a way to accomplish this without an explicit `deref` anywhere... but without deref-patterns, it seems there isn't.